From bb03b28922384aa30b08972feb95f6590092a339 Mon Sep 17 00:00:00 2001
From: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
Date: Thu, 13 Jun 2013 14:47:16 +0900
Subject: [PATCH 533/715] ARM: shmobile: r8a7790: Add scu_pcm driver support

Signed-off-by: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
(cherry picked from commit ca23d1eec32c89753516e6605db30cb9dfed2096)

Conflicts:
	arch/arm/mach-shmobile/clock-r8a7790.c
	arch/arm/mach-shmobile/setup-r8a7790.c

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 arch/arm/configs/shmobile_defconfig    |    2 ++
 arch/arm/mach-shmobile/board-lager.c   |    7 ++++++
 arch/arm/mach-shmobile/clock-r8a7790.c |    9 +++++++
 arch/arm/mach-shmobile/setup-r8a7790.c |   42 ++++++++++++++++++++++++++++++++
 4 files changed, 60 insertions(+)

diff --git a/arch/arm/configs/shmobile_defconfig b/arch/arm/configs/shmobile_defconfig
index 46e4ba3..be0978d 100644
--- a/arch/arm/configs/shmobile_defconfig
+++ b/arch/arm/configs/shmobile_defconfig
@@ -143,6 +143,8 @@ CONFIG_LOGO=y
 # CONFIG_SND_DRIVERS is not set
 # CONFIG_SND_ARM is not set
 CONFIG_SND_SOC_SH4_FSI=y
+CONFIG_SND_SOC_SCU=y
+CONFIG_SND_SCU_LAGER=y
 CONFIG_USB=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_EHCI_ROOT_HUB_TT=y
diff --git a/arch/arm/mach-shmobile/board-lager.c b/arch/arm/mach-shmobile/board-lager.c
index 5e3760e..69139ed 100644
--- a/arch/arm/mach-shmobile/board-lager.c
+++ b/arch/arm/mach-shmobile/board-lager.c
@@ -24,11 +24,16 @@
 #include <linux/kernel.h>
 #include <linux/pinctrl/machine.h>
 #include <linux/platform_device.h>
+#include <linux/i2c.h>
 #include <mach/common.h>
 #include <mach/r8a7790.h>
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
 
+static struct i2c_board_info lager_i2c_devices[] = {
+	{ I2C_BOARD_INFO("ak4642", 0x12), },
+};
+
 static const struct pinctrl_map lager_pinctrl_map[] = {
 	/* SCIF0 (CN19: DEBUG SERIAL0) */
 	PIN_MAP_MUX_GROUP_DEFAULT("sh-sci.6", "pfc-r8a7790",
@@ -65,6 +70,8 @@ static void __init lager_add_standard_devices(void)
 	r8a7790_pinmux_init();
 
 	r8a7790_add_standard_devices();
+	i2c_register_board_info(2, lager_i2c_devices,
+				ARRAY_SIZE(lager_i2c_devices));
 }
 
 static const char *lager_boards_compat_dt[] __initdata = {
diff --git a/arch/arm/mach-shmobile/clock-r8a7790.c b/arch/arm/mach-shmobile/clock-r8a7790.c
index ca0a8b6..8235a17 100644
--- a/arch/arm/mach-shmobile/clock-r8a7790.c
+++ b/arch/arm/mach-shmobile/clock-r8a7790.c
@@ -272,6 +272,15 @@ static struct clk_lookup lookups[] = {
 	CLKDEV_DEV_ID("ssi0", &mstp_clks[MSTP1015]),
 	CLKDEV_DEV_ID("ssi1", &mstp_clks[MSTP1014]),
 	CLKDEV_DEV_ID("ssi", &mstp_clks[MSTP1005]),
+	CLKDEV_CON_ID("adg", &mstp_clks[MSTP922]),
+	CLKDEV_CON_ID("src0", &mstp_clks[MSTP1031]),
+	CLKDEV_CON_ID("src1", &mstp_clks[MSTP1030]),
+	CLKDEV_CON_ID("dvc0", &mstp_clks[MSTP1019]),
+	CLKDEV_CON_ID("dvc1", &mstp_clks[MSTP1018]),
+	CLKDEV_CON_ID("scu", &mstp_clks[MSTP1017]),
+	CLKDEV_CON_ID("ssi0", &mstp_clks[MSTP1015]),
+	CLKDEV_CON_ID("ssi1", &mstp_clks[MSTP1014]),
+	CLKDEV_CON_ID("ssi", &mstp_clks[MSTP1005]),
 };
 
 static void __init r8a7790_rgx_control_init(void)
diff --git a/arch/arm/mach-shmobile/setup-r8a7790.c b/arch/arm/mach-shmobile/setup-r8a7790.c
index 624e595..e5f583f 100644
--- a/arch/arm/mach-shmobile/setup-r8a7790.c
+++ b/arch/arm/mach-shmobile/setup-r8a7790.c
@@ -1298,6 +1298,46 @@ static struct platform_device sysdmau_device = {
 	},
 };
 
+/* Audio */
+static struct platform_device alsa_soc_platform_device = {
+	.name		= "lager_alsa_soc_platform",
+	.id		= 0,
+};
+
+static struct resource scu_resources[] = {
+	[0] = {
+		.name   = "scu",
+		.start  = 0xec000000,
+		.end    = 0xec500fff,
+		.flags  = IORESOURCE_MEM,
+	},
+	[1] = {
+		.name   = "ssiu",
+		.start  = 0xec540000,
+		.end    = 0xec54085f,
+		.flags  = IORESOURCE_MEM,
+	},
+	[2] = {
+		.name   = "ssi",
+		.start  = 0xec541000,
+		.end    = 0xec54127f,
+		.flags  = IORESOURCE_MEM,
+	},
+	[3] = {
+		.name   = "adg",
+		.start  = 0xec5a0000,
+		.end    = 0xec5a0067,
+		.flags  = IORESOURCE_MEM,
+	},
+};
+
+static struct platform_device scu_device = {
+	.name		= "scu-pcm-audio",
+	.id		= 0,
+	.num_resources	= ARRAY_SIZE(scu_resources),
+	.resource	= scu_resources,
+};
+
 static struct platform_device *r8a7790_early_devices[] __initdata = {
 	&eth_device,
 	&powervr_device,
@@ -1321,6 +1361,8 @@ static struct platform_device *r8a7790_early_devices[] __initdata = {
 	&audmau_device,
 	&audmapp_device,
 	&sysdmau_device,
+	&alsa_soc_platform_device,
+	&scu_device,
 };
 
 static struct renesas_irqc_config irqc0_data = {
-- 
1.7.10.4

